# File: 05.1_salmon_array_job.R
# Auth: umar.niazi@kcl.as.uk
# DESC: create a parameter file and shell script to run array job on hpc
# Date: 14/9/2020


## set variables and source libraries
source('header.R')

## connect to mysql database to get sample information
library('RMySQL')

##### connect to mysql database to get samples
db = dbConnect(MySQL(), user='rstudio', password='12345', dbname='Projects', host='127.0.0.1')
dbListTables(db)

# check how many files each sample has
g_did
q = paste0('select count(File.idSample) as files, Sample.idData, Sample.title, Sample.id as SampleID , Sample.group1 as g1 from File, Sample
           where (Sample.idData = 49 and File.idSample = Sample.id and File.type like "fastq") group by File.idSample')
dfQuery = dbGetQuery(db, q)
dfQuery$title = gsub(" ", "", dfQuery$title, fixed = T)
dim(dfQuery)
dfQuery = na.omit(dfQuery)
head(dfQuery)

# get the count of files
q = paste0('select File.*, Sample.idData, Sample.group1 as g1 from File, Sample 
           where (Sample.idData = 49) and (File.idSample = Sample.id) and (File.type like "fastq")')
dfCounts = dbGetQuery(db, q)
dfCounts$name = gsub(" ", "", dfCounts$name, fixed = T)
head(dfCounts)
nrow(dfCounts)
dim(dfCounts)

# for each sample id, get the corresponding files
cvQueries = paste0('select File.*, Sample.title from File, Sample 
                   where (Sample.idData = 49 and Sample.id =', dfQuery$SampleID, ') and (File.idSample = Sample.id) and File.type like "fastq" ')

# set header variables 
cvShell = '#!/bin/bash -l'
cvJobName = '#SBATCH --job-name=salmon-array'
cvNodes = '#SBATCH --nodes=1'
cvProcessors = '#SBATCH --ntasks=8'
# format d-hh:mm:ss
cvRuntime = '#SBATCH --time=1-05:005:00'
cvPartition = '#SBATCH --partition brc'
# How much memory you need.
# --mem will define memory per node and
# --mem-per-cpu will define memory per CPU/core. Choose one of those.
cvMemoryReserve = '#SBATCH --mem-per-cpu=6000MB'
# Turn on mail notification. There are many possible self-explaining values:
# NONE, BEGIN, END, FAIL, ALL (including all aforementioned)
# For more values, check "man sbatch"
cvMail = '#SBATCH --mail-type=END,FAIL'
# set array job loop
length(cvQueries)
dim(dfCounts)
cvArrayJob = '#SBATCH --array=1-60'

# set the directory names
cvInput = 'input/'
cvOutput = 'output/Salmon/'
cvSalmon = 'salmon'
cvGeneIndex = '/users/k1625253/scratch/old-scratch_rosalind-legacy-import_2020-01-28/Data/MetaData/GenomeIndex/salmon_transcriptome_index/gencode.v35.transcripts_index/'

# create a parameter file and shell script
dir.create('AutoScripts')
oFile.param = file('AutoScripts/salmon_param.txt', 'wt')

temp = sapply(cvQueries, function(x){
  # get the file names
  dfFiles = dbGetQuery(db, x)
  # check for null return
  if (nrow(dfFiles) == 0) return();
  # remove white space from title
  dfFiles$title = gsub(" ", "", dfFiles$title, fixed=T)
  # split the file names into paired end 1 and 2, identified by R1 and R2 in the file name
  f = dfFiles$name
  d = grepl('_R1_', f)
  d = as.character(d)
  d[d == 'TRUE'] = 'R1'
  d[d == 'FALSE'] = 'R2'
  lf = split(f, d)
  # write salmon input file names
  in.r1 = paste0(cvInput, 'trim_', lf[[1]])
  in.r2 = paste0(cvInput, 'trim_', lf[[2]])
  out.sam = paste0(cvOutput, lf[[1]], '_quant')
  p1 = paste(in.r1, in.r2, out.sam, sep=' ')
  writeLines(p1, oFile.param)
})

close(oFile.param)

oFile = file('AutoScripts/salmon.sh', 'wt')

writeLines(c(cvShell, cvJobName, cvArrayJob, cvNodes, cvProcessors, 
             cvRuntime, cvPartition, cvMemoryReserve, cvMail), oFile)

writeLines(c('# Autogenerated script from salmon_array_job.R', paste('# date', date())), oFile)
writeLines(c('# make sure directory paths exist before running script'), oFile)
writeLines('\n\n', oFile)

# module load
writeLines(c('module load apps/salmon/1.2.1-singularity'), oFile)
writeLines('\n\n', oFile)

## write array job lines
writeLines("# Parse parameter file to get variables.
number=$SLURM_ARRAY_TASK_ID
paramfile=salmon_param.txt
 
inr1=`sed -n ${number}p $paramfile | awk '{print $1}'`
inr2=`sed -n ${number}p $paramfile | awk '{print $2}'`
outsal=`sed -n ${number}p $paramfile | awk '{print $3}'`

# 9. Run the program.", oFile)
p1 = paste('salmon quant -i', cvGeneIndex, '-p 8 -l ISR', '-1', '$inr1', '-2', '$inr2', '-o', '$outsal', sep=' ')
com = paste(p1)

writeLines(com, oFile)
writeLines('\n\n', oFile)

close(oFile)
dbDisconnect(db)
